### PR TITLE
[SYCL] Exclude SYCL_EXTERNAL functions from device image with the option -support-dynamic-linking

### PR DESCRIPTION
**Pull request by LU-JOHN**
With the option -support-dynamic-linking, SYCL External functions are not added to a device image even if they satisfy a dependency.
This work is part of adding dynamic linking support for SYCL.
Design document: https://github.com/intel/llvm/blob/sycl/sycl/doc/design/SharedLibraries.md

### CODE REVIEW

**Comment by asudarsa** [comment]
Hi @LU-JOHN 

Thanks for the PR. Can you please update the PR description with some high level context? Link to the dynamic linking doc etc could be handy.

Thanks

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 1]
```diff
@@ -234,6 +234,12 @@ cl::opt<bool> EmitOnlyKernelsAsEntryPoints{
              "device code split"),
     cl::cat(PostLinkCat), cl::init(false)};
 
+cl::opt<bool> SplitForRunTimeLinking{
+    "split-for-run-time-linking",
+    cl::desc("Do not add SYCL_EXTERNAL functions that are called to device "
```
Comment (asudarsa) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:239]:
```suggestion
    cl::desc("Do not add SYCL_EXTERNAL functions that are called inside device "
```

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp [hunk 2]
```diff
@@ -1191,7 +1200,7 @@ splitSYCLModule(std::unique_ptr<Module> M, ModuleSplitterSettings Settings) {
   ModuleDesc MD = std::move(M); // makeModuleDesc() ?
   // FIXME: false arguments are temporary for now.
   auto Splitter =
-      getDeviceCodeSplitter(std::move(MD), Settings.Mode, false, false);
+      getDeviceCodeSplitter(std::move(MD), Settings.Mode, false, false, false);
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:1203]: could we add comments for what all the bool arguments are here? like
```
/*foo=*/false
```

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp [hunk 3]
```diff
@@ -177,6 +177,11 @@ class DependencyGraph {
 
     // We add every function into the graph
     for (const auto &F : M.functions()) {
+
+      if (SplitForRunTimeLinking &&
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:181]: The variable name and the behavior are a bit misaligned/nondescriptive to me. It seems like what the bool really controls is it we include SYCL_EXTERNAL functions into the image, maybe we could name the variable `IncludeExternalFunctions` or something

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 4]
```diff
@@ -234,6 +234,12 @@ cl::opt<bool> EmitOnlyKernelsAsEntryPoints{
              "device code split"),
     cl::cat(PostLinkCat), cl::init(false)};
 
+cl::opt<bool> SplitForRunTimeLinking{
+    "split-for-run-time-linking",
```
Comment (sarnex) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:238]: Do we also need to hook up the driver/clang-linker-wrapper passing this, or will that be a separate PR?

Comment (sarnex) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:238]: `for` is a bit confusing for me, maybe we can name this `link-external-functions-at-runtime` or something like that?

**Comment by asudarsa** [comment]
Adding @sergey-semenov for heads up. This change is part of the effort to add Dynamic Linking support for SYCL device images. When we run sycl-post-link using this option, the incoming graph will no longer be fully contained and SYCL runtime will need to react accordingly.
Currently, the option is not added and no changes in codegen by sycl-post-link is expected.

Thanks

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (LU-JOHN) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:238]: How about `include-external-functions` or `include-exernal-functions-in-device-split`?  The option and variable names should be very similar.

Comment (sarnex) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:238]: sounds good to me!

Comment (LU-JOHN) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:238]: Since this is a new option, let's invert the sense so that it is not required to get previous behavior, `exclude-external-functions`.

Comment (sarnex) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:238]: no strong opinion there

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp (hunk 2 above)
Comment (LU-JOHN) [reply llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:1203]: Added comments for bool args.

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp (hunk 3 above)
Comment (LU-JOHN) [reply llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:181]: Updated variable to ExcludeExternalFunctions.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (LU-JOHN) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:238]: Changed option to -exclude-external-functions

File: llvm/test/tools/sycl-post-link/exclude_external_functions.ll [hunk 5]
```diff
@@ -0,0 +1,67 @@
+; This test checks that when device code splitting is done for run-time linking,
+; dependencies to SYCL External functions do not cause the SYCL External function
+; to be added to a device image.
+
+; RUN: sycl-post-link -symbols -exclude-external-functions -split=kernel -S < %s -o %t.table
+
+
+; RUN: FileCheck %s -input-file=%t_0.sym --check-prefixes CHECK-SYM-0
+; RUN: FileCheck %s -input-file=%t_1.sym --check-prefixes CHECK-SYM-1
```
Comment (sarnex) [inline llvm/test/tools/sycl-post-link/exclude_external_functions.ll:9]: it seems we have two LIT tests but they're exactly the same minus a comment, do we want to delete one of them?

Comment (LU-JOHN) [reply llvm/test/tools/sycl-post-link/exclude_external_functions.ll:9]: Deleted poorly named file.

**Review by sarnex — APPROVED** [review]

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 1 above)
Comment (LU-JOHN) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Updated description.

File: llvm/test/tools/sycl-post-link/exclude_external_functions.ll [hunk 6]
```diff
@@ -0,0 +1,67 @@
+; Test that when the -exclude-exernal-funtions option is used,
+; dependencies to SYCL External functions do not cause the SYCL External function
+; to be added to a device image.
+
+; RUN: sycl-post-link -symbols -exclude-external-functions -split=kernel -S < %s -o %t.table
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/exclude_external_functions.ll:5]: I think that we should add a separate `per-source` split test that contains calls between `SYCL_EXTERNAL` functions from both different and the same translation units.

In that device code split mode, we don't expect `SYCL_EXTERNAL` functions that belong to the same translation unit to be outlined into different device images regardless of `-exclude-external-functions` flag presence.

Comment (LU-JOHN) [reply llvm/test/tools/sycl-post-link/exclude_external_functions.ll:5]: New test added, exclude_external_functions_source.ll

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 4 above)
Comment (LU-JOHN) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Changes made to driver to pass -exclude-external-functions if -shared seen.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 7]
```diff
@@ -10686,6 +10686,9 @@ getTripleBasedSYCLPostLinkOpts(const ToolChain &TC, const JobAction &JA,
       (NewOffloadDriver || (!Triple.isNVPTX() && !Triple.isAMDGPU())))
     addArgs(PostLinkArgs, TCArgs, {"-emit-only-kernels-as-entry-points"});
 
+  if (TCArgs.hasArg(options::OPT_shared))
+    addArgs(PostLinkArgs, TCArgs, {"-exclude-external-functions"});
+
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:10691]: Please move to the `getOtherSYCLPostLinkOpts` function.  It doesn't look like this option is dependent on any triple behaviors.

Comment (asudarsa) [reply clang/lib/Driver/ToolChains/Clang.cpp:10691]: Nit: Think I call it getNonTripleBasedSYCLPostLinkOpts now. :-)

Comment (LU-JOHN) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Moved change to getNonTripleBasedSYCLPostLinkOpts.

File: llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h [hunk 8]
```diff
@@ -261,11 +261,13 @@ class ModuleSplitterBase {
...
 SmallVector<ModuleDesc, 2> splitByESIMD(ModuleDesc &&MD,
-                                        bool EmitOnlyKernelsAsEntryPoints);
+                                        bool EmitOnlyKernelsAsEntryPoints,
+                                        bool ExcludeExternalFunctions);
 
 std::unique_ptr<ModuleSplitterBase>
 getDeviceCodeSplitter(ModuleDesc &&MD, IRSplitMode Mode, bool IROutputOnly,
-                      bool EmitOnlyKernelsAsEntryPoints);
+                      bool EmitOnlyKernelsAsEntryPoints,
+                      bool ExcludeExternalFunctions);
```
Comment (asudarsa) [inline llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h:270]: Sorry for catching this late. Think this sounds better as ExcludeSYCLExternalFunctions?

Thanks

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks. One nit. Not a blocker.

**Review by AlexeySachkov — APPROVED** [review]

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 9]
```diff
@@ -10647,8 +10647,11 @@ static void getNonTripleBasedSYCLPostLinkOpts(const ToolChain &TC,
                      options::OPT_fsycl_esimd_force_stateless_mem, false))
     addArgs(PostLinkArgs, TCArgs, {"-lower-esimd-force-stateless-mem=false"});
 
+  // TODO Enable when SYCL RT supports dynamic linking
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:10650]: It will be cleaner and also avoid possible merge conflicts to remove this block of code and reintroduce when support is ready. Tracking can be done elsewhere. Thanks

File: clang/test/Driver/sycl-offload.cpp [hunk 10]
```diff
@@ -179,10 +179,11 @@
 // CHECK_SYCL_POST_LINK_OPT_NO_PASS-NOT: sycl-post-link{{.*}}emit-only-kernels-as-entry-points
 
 /// Check selective passing of -support-dynamic-linking to sycl-post-link tool
-// RUN: %clang -### -target x86_64-unknown-linux-gnu -fsycl -fsycl-targets=spir64_fpga -shared %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_SHARED_PASS %s
-// RUN: %clang -### -target x86_64-unknown-linux-gnu -fsycl -fsycl-targets=spir64_gen -shared %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_SHARED_PASS %s
+// Enable when SYCL RT supports dynamic linking
+// TODO: %clang -### -target x86_64-unknown-linux-gnu -fsycl -fsycl-targets=spir64_fpga -shared %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_SHARED_PASS %s
+// TODO: %clang -### -target x86_64-unknown-linux-gnu -fsycl -fsycl-targets=spir64_gen -shared %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_SHARED_PASS %s
```
Comment (asudarsa) [inline clang/test/Driver/sycl-offload.cpp:184]: Please follow usual conventions. 
182 // TODO: Enable when SYCL RT supports dynamic linking.
183 // RUNx: %clang
184 // RUNx: %clang
185
186 // RUNx: %clang

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 9 above)
Comment (LU-JOHN) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Code removed.

**Comment by asudarsa** [comment]
What is the reasoning behind the most recent commit? Thanks

**Comment by LU-JOHN** [comment]
> What is the reasoning behind the most recent commit? Thanks

Thanks for the review.  

The test case ESIMD/regression/complex-lib-lin.cpp failed because the symbols "__itt\*" were split into another device image, and the test failed because the symbols were not found.  As far as I can tell these symbols will be found once the SYCL RT supports dynamic linking, but @AlexeySachkov recommended not allowing symbols starting with "\_\_" as dependencies.

@AlexeySachkov, can you comment on this?  Thanks.

File: clang/test/Driver/sycl-offload.cpp [hunk 11]
```diff
@@ -178,6 +178,14 @@
 // RUN: %clang -### -target x86_64-unknown-linux-gnu -fsycl -fsycl-targets=spir64_gen -fno-sycl-remove-unused-external-funcs %s 2>&1 | FileCheck -check-prefix=CHECK_SYCL_POST_LINK_OPT_NO_PASS %s
 // CHECK_SYCL_POST_LINK_OPT_NO_PASS-NOT: sycl-post-link{{.*}}emit-only-kernels-as-entry-points
 
+/// Check selective passing of -support-dynamic-linking to sycl-post-link tool
```
Comment (asudarsa) [inline clang/test/Driver/sycl-offload.cpp:181]: Do we need to disable this test as we have removed the support in driver code?

**Comment by asudarsa** [comment]
I expect tweaking of 'canBeImportedFunction' as we do further testing with the -supports-dynamic-linking option turned on for '-shared' compilation.
Also, @LU-JOHN, please update document with how sycl-post-link works as of now.

Thanks all for reviews.

**Comment by LU-JOHN** [comment]
@intel/llvm-gatekeepers 
Failure in post-commit https://github.com/intel/llvm/actions/runs/9665746765/job/26663614981 looks like an infrastructure issue.     Log shows the test reports results are correct:

```
# .---command stdout------------
# | The results are correct!
# `-----------------------------
```

but non-zero exit status is returned.
